File: tools/editor/shared/rule_knowledge_validation.py
"""Mirror of src/core/ruleKnowledgeValidation.ts; shared-case tests enforce parity."""


def validate_rule_knowledge(rule, graphs, fragments=()):
    owned = [g for g in graphs if g.get('ownerType') == 'rule' and g.get('ownerId') == rule.get('id')]
    if 'narrativeStates' not in rule:
        return ['rule.mode.missing'] if owned else []
    variants = rule['narrativeStates']
    if not isinstance(variants, dict) or not variants:
        return ['rule.states.invalid']
    errors = []
    if len(owned) != 1:
        errors.append('rule.owner.count')
    if any(g.get('run') is not None for g in owned):
        errors.append('rule.owner.repeatable')
    if any(f.get('ruleId') == rule.get('id') for f in fragments):
        errors.append('rule.fragments.mixed')
    states = list(owned[0].get('states', {})) if len(owned) == 1 else []
    for state in states:
        if state not in variants:
            errors.append(f'rule.state.missing:{state}')
    for state, variant in variants.items():
        if len(owned) == 1 and state not in states:
            errors.append(f'rule.state.unknown:{state}')
        if not isinstance(variant, dict) or not isinstance(variant.get('layers'), dict):
            errors.append(f'rule.layers.invalid:{state}')
            continue
        for layer, body in variant['layers'].items():
            if layer not in ('xiang', 'li', 'shu') or layer not in (rule.get('layers') or {}):
                errors.append(f'rule.layer.undefined:{state}:{layer}')
            if not isinstance(body, dict) or not isinstance(body.get('text'), str) or not body['text'].strip():
                errors.append(f'rule.text.empty:{state}:{layer}')
            if isinstance(body, dict) and 'verified' in body and body['verified'] not in ('unverified', 'effective', 'questionable'):
                errors.append(f'rule.verified.invalid:{state}:{layer}')
    return errors

File: src/core/ruleKnowledgeValidation.ts
import type { RuleDef, RuleFragmentDef } from '../data/types';

export interface RuleOwnerGraph {
  id: string;
  ownerType: string;
  ownerId?: string;
  states: Record<string, unknown>;
  run?: unknown;
}

/** Authoritative rule projection checks; Python mirrors these codes with shared cases. */
export function validateRuleKnowledge(
  rule: RuleDef, graphs: RuleOwnerGraph[], fragments: RuleFragmentDef[] = [],
): string[] {
  const owned = graphs.filter(g => g.ownerType === 'rule' && g.ownerId === rule.id);
  if (rule.narrativeStates === undefined) return owned.length ? ['rule.mode.missing'] : [];
  const errors: string[] = [];
  const variants = rule.narrativeStates;
  if (!variants || typeof variants !== 'object' || Array.isArray(variants) || !Object.keys(variants).length) {
    return ['rule.states.invalid'];
  }
  if (owned.length !== 1) errors.push('rule.owner.count');
  if (owned.some(g => g.run !== undefined && g.run !== null)) errors.push('rule.owner.repeatable');
  if (fragments.some(f => f.ruleId === rule.id)) errors.push('rule.fragments.mixed');
  const states = owned.length === 1 ? Object.keys(owned[0]!.states) : [];
  for (const state of states) {
    if (!(state in variants)) errors.push(`rule.state.missing:${state}`);
  }
  for (const [state, variant] of Object.entries(variants)) {
    if (owned.length === 1 && !states.includes(state)) errors.push(`rule.state.unknown:${state}`);
    if (!variant || typeof variant.layers !== 'object' || !variant.layers || Array.isArray(variant.layers)) {
      errors.push(`rule.layers.invalid:${state}`);
      continue;
    }
    for (const [layer, body] of Object.entries(variant.layers)) {
      if (!['xiang', 'li', 'shu'].includes(layer) || !Object.prototype.hasOwnProperty.call(rule.layers ?? {}, layer)) {
        errors.push(`rule.layer.undefined:${state}:${layer}`);
      }
      if (!body || typeof body.text !== 'string' || !body.text.trim()) errors.push(`rule.text.empty:${state}:${layer}`);
      if (body?.verified !== undefined && !['unverified', 'effective', 'questionable'].includes(body.verified)) {
        errors.push(`rule.verified.invalid:${state}:${layer}`);
      }
    }
  }
  return errors;
}
